lib/ansible/inventory/group.py

```diff
@@ -83,23 +86,38 @@ def __repr__(self):
     def __str__(self):
         return self.get_name()
 
+    def __hash__(self):
+        return id(self)
+
+    def __eq__(self, other):
+        return not self.__ne__(other)
+
+    def __ne__(self, other):
+        if not isinstance(other, Group):
+            return True
+
+        if self.serialize() != other.serialize():
+            return True
+
+        return False
+
     def __getstate__(self):
         return self.serialize()
 
     def __setstate__(self, data):
         return self.deserialize(data)
 
     def serialize(self):
-        parent_groups = []
-        for parent in self.parent_groups:
-            parent_groups.append(parent.serialize())
+        parent_groups = [g.name for g in self.parent_groups]
+        child_groups = [g.name for g in self.child_groups]
 
         self._hosts = None
```

Comment:
this should be `self.hosts` .. to avoid confusion i would rename `self._hosts` to `self._host_names`  as `self.hosts` is the object list and `self._hosts` is the 'host name' list

Reply:
There is a host_names property that uses self._hosts. Both self.host_names and self._hosts are sets so using them may have order implications. Modifying them might break external things using them. If it's the same to you, I'm going to leave those alone and add another list comprehension.

Reply:
Also setting self._hosts is fine here I *think* but not necessary. Resetting self.hosts would break things since we don't actually want to reset the current instance when we call serialize.